Analytics: Report unknown IOS requests

leoetlino opened the pull request:
We are now down to a few unknown requests for some resource managers (ES & USB), so having info about the remaining ones will help get them implemented.

Also report unknown IOS devices along with the current IOS version, as there may be devices we still haven't implemented.

Only for open, ioctl and ioctlv (the other types aren't too useful).

sepalani commented:
Any reason not to report the other IOS requests, for instance the Network ones (knowing that we even log some of them)?

BhaaLseN commented:
As mentioned on IRC, it's probably not necessary to include the full base- and pergame-builder stuff; we have that with the other requests anyways. Just creating an empty builder should help? Except that we'd need to add gameid and titleid elsewhere (perhaps in a smaller/similar variant of `MakePerGameBuilder`  that only adds those?)
I don't quite know why `MakePerGameBuilder` and `m_per_game_builder` are there, but I don't see a reason to have the contents being copied to multiple requests from the same client (especially for those where we might get multiple reports from the same client within the same game/title).

Helios747 commented:
Thinking about UI stuff, will we need to modify what we tell the user on first launch?

If we're just logging unknown commands, I don't see the immediate need to update the text. No actual data is being recorded as far as I can tell.

Will this change down the road?

leoetlino commented:
@sepalani my initial plan was to only have reports for mostly complete resource managers. But it's a better idea to just try and catch any unknown request. So I've added reports for all RMs where we don't know if the ioctl/ioctlv list is 100% complete.

@Helios747 I don't think this will change later. Analytics will never be used to record personal user data anyway.

sepalani commented on Source/Core/Core/IOS/Network/SSL.cpp:
```diff
@@ -541,6 +542,7 @@ IPCCommandResult NetSSL::IOCtlV(const IOCtlVRequest& request)
   }
   default:
     request.DumpUnknown(GetDeviceName(), LogTypes::IOS_SSL);
+    DolphinAnalytics::Instance()->ReportUnknownIOSRequest(GetDeviceName(), request);
```
There is a missing break.

leoetlino replied:
Added.

sepalani commented on Source/Core/Core/IOS/USB/OH0/OH0.cpp:
```diff
@@ -311,6 +314,7 @@ IPCCommandResult OH0::DeviceIOCtlV(const u64 device_id, const IOCtlVRequest& req
+    DolphinAnalytics::Instance()->ReportUnknownIOSRequest(GetDeviceName(), request);
     return GetDefaultReply(IPC_SUCCESS);
   default:
     return GetDefaultReply(IPC_EINVAL);
```
Is there a reason not to report the default case? Ditto for DeviceIOCtl.

BhaaLseN replied:
Asked that when the code was initially written; its because there are no other known IOCTLs beyond the ones in the enum - the default case should never be hit in practise (other than in broken homebrew)

delroth commented:
@Helios747 this use case was planned for when I initially designed the analytics system (that's the main reason why it exists -- the usage stats is secondary). So I think everything privacy related should already be fine for that.

@leoetlino my main problem with this PR is that it does not implement any kind of report deduplication on the client side (as far as I can tell). What if a game is blasting unknown ioctls e.g. once per frame in the background? I'd prefer if we could limit this to one per {game run, device name, ioctl type} via a locally maintained unordered_set.

I'm also not convinced we need to log argument sizes. I fail to see how it will be useful in terms of data collection. Once we have the game and the ioctl type, we will anyway do some in-game testing to implement the missing ioctls... and then we'll get all the size info we want. The main reason why I care is because it makes the event deduplication more annoying.

delroth commented on Source/Core/Core/Analytics.cpp:
```diff
@@ -116,6 +118,61 @@ void DolphinAnalytics::ReportGameStart()
+  Send(builder);
+}
+
+void DolphinAnalytics::ReportUnknownIOSRequest(const std::string& path,
```
One thing design-wise I'm not completely convinced about: intuitively, I think these report building functions should be defined near the IOS code, not near the analytics code. This is because otherwise we might end up growing too many dependencies from analytics to the rest of the codebase. I'd rather have the dependencies be in the other direction: random modules using analytics -> analytics.

But again: I'm not set on any of the two, I'd just like to make sure you considered the alternative.

leoetlino replied:
I put it there to be somewhat consistent with ReportGameStart. However, it's true that this adds a dependency from analytics to IOS... Would you prefer if this is moved to near the IOS code instead of being here? (I don't really mind either way.)

leoetlino commented:
@delroth Right, event deduplication isn't implemented in this PR. I wasn't sure how to do it in a generic way (as suggested on IRC). For these IOS reports, a unordered_set with {device name, request type, ioctl number} would probably be enough indeed. I'll add this in a moment.

delroth commented:
Yeah, I think I would prefer it that way (functions near the IOS code).
On Sat, 15 Apr 2017 at 17:56, Léo Lam <notifications@github.com> wrote:

> *@leoetlino* commented on this pull request.
> ------------------------------
>
> In Source/Core/Core/Analytics.cpp
> <https://github.com/dolphin-emu/dolphin/pull/5264#discussion_r111661494>:
>
> > @@ -116,6 +118,61 @@ void DolphinAnalytics::ReportGameStart()
>    Send(builder);
>  }
>
> +void DolphinAnalytics::ReportUnknownIOSDevice(const std::string& path)
> +{
> +  Common::AnalyticsReportBuilder builder = MakeBuilderWithGameID();
> +  builder.AddData("type", "ios-unknown-device");
> +  builder.AddData("ios-version", IOS::HLE::GetVersion());
> +  builder.AddData("ios-path", path);
> +  Send(builder);
> +}
> +
> +void DolphinAnalytics::ReportUnknownIOSRequest(const std::string& path,
>
> I put it there to be somewhat consistent with ReportGameStart. However,
> it's true that this adds a dependency from analytics to IOS... Would you
> prefer if this is moved to near the IOS code instead of being here? (I
> don't really mind either way.)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dolphin-emu/dolphin/pull/5264#discussion_r111661494>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMYLn3z2pLkLxhMDQPVevohCegqIZZAks5rwIYzgaJpZM4M906h>
> .
>

leoetlino commented:
I've moved the report building functions to Core/IOS. (Note to self: IOS/IPC should probably be split at some point.) I've also removed the ioctlv vector sizes -- but kept the counts because there are ioctlvs which have different behaviour depending on the number of vectors.

For tracking already sent reports, do you have any suggestion for doing this in a more generic way? Since we'll probably want to deduplicate unknown device reports and other analytics reports... I've been thinking about keeping hashes of reports but I'm pretty sure there is a better way.

delroth commented:
I have no good suggestion for deduplication at this time. Happy to review designs though, I'm on vacation so the last thing I want to do is to start spending hours designing this :)

shuffle2 commented:
For deduplication it should be enough to keep a list of already-sent event keys in a local cache, and only send up new things on successive attempts? A full dedupe system would involve the client prefetching this cache from the server, so the server can tell clients to ignore events which other clients already reported (but I really doubt we'll do that :p).

It's sad that we don't have some good local format to do this in (afaik?) - the Ini code would work but this is an example why we really should have moved away from Ini already.

delroth commented:
I don't care about deduplicating across multiple clients or across multiple
runs or Dolphin, I just want to make it REALLY hard for us to have a
situation where a client is sending us tens of events per second because
someone thought a condition was rare and it in fact is not. Some random in
memory thing is way more than enough.

On Tue, Jun 6, 2017 at 5:47 AM shuffle2 <notifications@github.com> wrote:

> For deduplication it should be enough to keep a list of already-sent event
> keys in a local cache, and only send up new things on successive attempts?
> A full dedupe system would involve the client prefetching this cache from
> the server, so the server can tell clients to ignore events which other
> clients already reported (but I really doubt we'll do that :p).
>
> It's sad that we don't have some good local format to do this in (afaik?)
> - the Ini code would work but this is an example why we really should have
> moved away from Ini already.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dolphin-emu/dolphin/pull/5264#issuecomment-306373986>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMYLrMtqXeEG9URIGjjSF40QsfpCuoKks5sBMvngaJpZM4M906h>
> .
>

leoetlino commented:
Closing for now; 7 months later after this PR was opened, I don't think there's a lot of unimplemented IOS functionality we don't already know about.
